module-system/sov-state/src/scratchpad.rs

```diff
@@ -40,73 +41,114 @@ impl<S: Storage> Debug for RevertableDelta<S> {
 }
 
 /// A read-write set which can be committed as a unit
-pub enum WorkingSet<S: Storage> {
+enum ReadWriteSet<S: Storage> {
     Standard(Delta<S>),
     Revertable(RevertableDelta<S>),
 }
 
+/// This structure holds the read-write set and the events gathered during the execution of a transaction.
+pub struct WorkingSet<S: Storage> {
+    read_write_set: ReadWriteSet<S>,
+    events: Vec<Event>,
+}
+
 impl<S: Storage> WorkingSet<S> {
     pub fn new(inner: S) -> Self {
-        Self::Standard(Delta::new(inner))
+        Self {
+            read_write_set: ReadWriteSet::Standard(Delta::new(inner)),
+            events: Default::default(),
+        }
     }
 
     pub fn with_witness(inner: S, witness: S::Witness) -> Self {
-        Self::Standard(Delta::with_witness(inner, witness))
+        Self {
+            read_write_set: ReadWriteSet::Standard(Delta::with_witness(inner, witness)),
+            events: Default::default(),
+        }
     }
 
     pub fn to_revertable(self) -> Self {
-        match self {
-            WorkingSet::Standard(delta) => WorkingSet::Revertable(delta.get_revertable_wrapper()),
-            WorkingSet::Revertable(_) => self,
+        let read_write_set = match self.read_write_set {
+            ReadWriteSet::Standard(delta) => {
+                ReadWriteSet::Revertable(delta.get_revertable_wrapper())
+            }
+            ReadWriteSet::Revertable(_) => self.read_write_set,
+        };
+
+        Self {
+            read_write_set,
+            events: self.events,
         }
     }
 
     pub fn commit(self) -> Self {
-        match self {
-            s @ WorkingSet::Standard(_) => s,
-            WorkingSet::Revertable(revertable) => WorkingSet::Standard(revertable.commit()),
+        let read_write_set = match self.read_write_set {
+            s @ ReadWriteSet::Standard(_) => s,
+            ReadWriteSet::Revertable(revertable) => ReadWriteSet::Standard(revertable.commit()),
+        };
+
+        Self {
+            read_write_set,
+            events: self.events,
         }
     }
 
     pub fn revert(self) -> Self {
-        match self {
-            s @ WorkingSet::Standard(_) => s,
-            WorkingSet::Revertable(revertable) => WorkingSet::Standard(revertable.revert()),
+        let read_write_set = match self.read_write_set {
+            s @ ReadWriteSet::Standard(_) => s,
+            ReadWriteSet::Revertable(revertable) => ReadWriteSet::Standard(revertable.revert()),
+        };
+        Self {
+            read_write_set,
+            // The `revert` removes all events associated with the transaction
+            events: Vec::default(),
```

Comment:
I don't think this logic matches the state revert logic. To be consistent, don't we want to remove the events that have happened since the last call to `to_revertable()`?

Reply:
Synced with @bkolad offline. This divergence is intentional, since `events` are a per-transaction concept while state changes are accumulated per-batch.